File: sdks/typescript/packages/core/src/adapters/manual.ts
import type { Trace } from "../proto/types.js";
import { TraceBuilder } from "../trace.js";

export class ManualAdapter {
  private readonly agentId: string | undefined;

  constructor(agentId?: string) {
    this.agentId = agentId;
  }

  capture(builderFn: (builder: TraceBuilder) => void): Trace {
    const builder = new TraceBuilder(this.agentId);
    builderFn(builder);
    return builder.build();
  }

  createBuilder(): TraceBuilder {
    return new TraceBuilder(this.agentId);
  }
}

File: sdks/python/src/attest/adapters/manual.py
"""Manual trace adapter using TraceBuilder."""

from __future__ import annotations

from collections.abc import Callable

from attest._proto.types import Trace
from attest.adapters._base import BaseAdapter
from attest.trace import TraceBuilder


class ManualAdapter(BaseAdapter):
    """Adapter for manually constructing traces via TraceBuilder."""

    def capture(self, builder_fn: Callable[[TraceBuilder], None]) -> Trace:
        """Execute builder_fn with a TraceBuilder and return the built Trace.

        builder_fn receives a TraceBuilder and should call methods on it
        to construct the trace. It does not need to call build().
        """
        builder = self._create_builder()
        builder_fn(builder)
        return builder.build()

    def create_builder(self) -> TraceBuilder:
        """Create a new TraceBuilder for manual construction."""
        return self._create_builder()
